File: packages/core/src/utils/info-gap-ledger.ts
/**
 * 信息差账本（G7a/341 号，Phase B 批次二；WNW W5 信息差登记 + 泄密机检的采纳）。
 *
 * truth 类目 `story/info_gaps.md`（作者可手改，每条一节）：
 *
 *   ## gap-1
 *   - secret: 主角身世是前朝皇室
 *   - knows: 林动, 王胖
 *   - readerKnows: false
 *   - registeredAt: 3
 *   - keywords: 皇室, 前朝, 血脉

 * 机检两枚（确定性、无 LLM，供 continuity 审计维度 33/34 的本地预扫描）：
 * - **泄密机检** `detectSecretLeaks`：未到读者已知时点的秘密关键词出现在正文中，
 *   命中窗口含对白引号则升级（对白/内心戏泄密风险最高）；
 * - **废笔机检** `detectReaderRedundancy`：读者已知的秘密关键词再次复述——向读者
 *   复述已知信息属废笔。
 *
 * 双端：`engine-rs/src/utils/info_gap_ledger.rs` 1:1 镜像；共享向量
 * `src/__tests__/golden/info-gap-vectors.json`。
 */

export interface InfoGapEntry {
  readonly id: string;
  readonly secret: string;
  /** 知情角色（逗号/顿号分隔存储）。 */
  readonly knows: ReadonlyArray<string>;
  /** 读者是否已知情（true 后再复述即废笔）。 */
  readonly readerKnows: boolean;
  /** 登记章（秘密埋设的章节号）。 */
  readonly registeredAt: number;
  readonly keywords: ReadonlyArray<string>;
}

const GAP_HEADING = /^##\s+(gap-[\w-]+)\s*$/;
const FIELD_LINE = /^-\s*(\w+)\s*:\s*(.*)$/;

/** 解析信息差账本 markdown（作者手改容错：缺字段用空值，坏节跳过）。 */
export function parseInfoGapsMarkdown(markdown: string): InfoGapEntry[] {
  const entries: InfoGapEntry[] = [];
  let current: { id: string; fields: Map<string, string> } | undefined;
  const flush = () => {
    if (!current) return;
    const secret = current.fields.get("secret")?.trim() ?? "";
    const knows = splitList(current.fields.get("knows") ?? "");
    const readerKnows = /^(true|是|yes|1)$/i.test((current.fields.get("readerKnows") ?? "").trim());
    const registeredAt = Number.parseInt((current.fields.get("registeredAt") ?? "").trim(), 10);
    const keywords = splitList(current.fields.get("keywords") ?? "");
    if (secret || keywords.length > 0) {
      entries.push({
        id: current.id,
        secret,
        knows,
        readerKnows,
        registeredAt: Number.isFinite(registeredAt) ? registeredAt : 0,
        keywords,
      });
    }
    current = undefined;
  };

  for (const rawLine of markdown.split(/\r?\n/)) {
    const line = rawLine.trim();
    const heading = GAP_HEADING.exec(line);
    if (heading) {
      flush();
      current = { id: heading[1]!, fields: new Map() };
      continue;
    }
    if (!current) continue;
    const field = FIELD_LINE.exec(line);
    if (field) current.fields.set(field[1]!, field[2] ?? "");
  }
  flush();
  return entries;
}

function splitList(value: string): string[] {
  return value
    .split(/[,，、;；]/)
    .map((item) => item.trim())
    .filter(Boolean);
}

/** 渲染信息差账本（防呆方言：每条一节、平铺 key: value 行）。 */
export function renderInfoGapsMarkdown(entries: ReadonlyArray<InfoGapEntry>): string {
  const sections = entries.map((entry) =>
    [
      `## ${entry.id}`,
      `- secret: ${entry.secret}`,
      `- knows: ${entry.knows.join(", ")}`,
      `- readerKnows: ${entry.readerKnows}`,
      `- registeredAt: ${entry.registeredAt}`,
      `- keywords: ${entry.keywords.join(", ")}`,
    ].join("\n"),
  );
  return ["# 信息差账本（Info Gaps）", ...sections].join("\n\n");
}

// ── 机检 ──

const DIALOGUE_MARKS = /[「」『』“”‘’"']/;

export interface SecretLeakHit {
  readonly gapId: string;
  readonly keyword: string;
  readonly occurrences: number;
  /** 命中窗口含对白引号（对白/内心泄密风险最高）。 */
  readonly inDialogue: boolean;
  readonly knows: ReadonlyArray<string>;
}

/**
 * 泄密机检：readerKnows=false 的秘密关键词出现在正文中即候选；
 * 命中位置前后 24 字窗口含对白引号 → inDialogue=true（升级为必须审查）。
 * 同 gap 多关键词按命中数聚合，取最高危窗口判定 inDialogue。
 */
export function detectSecretLeaks(params: {
  readonly content: string;
  readonly gaps: ReadonlyArray<InfoGapEntry>;
  readonly currentChapter: number;
}): SecretLeakHit[] {
  const hits: SecretLeakHit[] = [];
  for (const gap of params.gaps) {
    if (gap.readerKnows) continue;
    if (gap.registeredAt > params.currentChapter) continue; // 未登记（未来秘密）不扫
    for (const keyword of gap.keywords) {
      if (!keyword) continue;
      const occurrences = countOccurrences(params.content, keyword);
      if (occurrences === 0) continue;
      const inDialogue = windowHasDialogue(params.content, keyword);
      hits.push({ gapId: gap.id, keyword, occurrences, inDialogue, knows: [...gap.knows] });
    }
  }
  return hits.sort(
    (a, b) =>
      Number(b.inDialogue) - Number(a.inDialogue)
      || b.occurrences - a.occurrences
      || a.gapId.localeCompare(b.gapId)
      || a.keyword.localeCompare(b.keyword),
  );
}

export interface ReaderRedundancyHit {
  readonly gapId: string;
  readonly keyword: string;
  readonly occurrences: number;
}

/** 废笔机检：读者已知的秘密关键词再次出现——向读者复述已知信息。 */
export function detectReaderRedundancy(params: {
  readonly content: string;
  readonly gaps: ReadonlyArray<InfoGapEntry>;
}): ReaderRedundancyHit[] {
  const hits: ReaderRedundancyHit[] = [];
  for (const gap of params.gaps) {
    if (!gap.readerKnows) continue;
    for (const keyword of gap.keywords) {
      if (!keyword) continue;
      const occurrences = countOccurrences(params.content, keyword);
      if (occurrences > 0) {
        hits.push({ gapId: gap.id, keyword, occurrences });
      }
    }
  }
  return hits.sort(
    (a, b) =>
      b.occurrences - a.occurrences
      || comparePlain(a.gapId, b.gapId)
      || comparePlain(a.keyword, b.keyword),
  );
}

/** 纯码元比较（禁 localeCompare——排序必须跨端/跨环境确定）。 */
function comparePlain(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

function countOccurrences(content: string, keyword: string): number {
  if (!keyword) return 0;
  const lower = content.toLowerCase();
  const needle = keyword.toLowerCase();
  let count = 0;
  let index = lower.indexOf(needle);
  while (index >= 0) {
    count += 1;
    index = lower.indexOf(needle, index + needle.length);
  }
  return count;
}

function windowHasDialogue(content: string, keyword: string, window = 24): boolean {
  const lower = content.toLowerCase();
  const needle = keyword.toLowerCase();
  let index = lower.indexOf(needle);
  while (index >= 0) {
    const from = Math.max(0, index - window);
    const to = Math.min(content.length, index + needle.length + window);
    if (DIALOGUE_MARKS.test(content.slice(from, to))) return true;
    index = lower.indexOf(needle, index + needle.length);
  }
  return false;
}

/** 审计注入文本：continuity 维度 33/34 的本地机检摘要（无命中返回 undefined）。 */
export function renderInfoGapAuditNotes(params: {
  readonly leaks: ReadonlyArray<SecretLeakHit>;
  readonly redundancies: ReadonlyArray<ReaderRedundancyHit>;
  readonly language?: "zh" | "en";
}): string | undefined {
  const language = params.language ?? "zh";
  if (params.leaks.length === 0 && params.redundancies.length === 0) return undefined;
  const lines: string[] = [];
  if (params.leaks.length > 0) {
    lines.push(
      language === "en"
        ? "Secret-leak pre-scan flagged (verify whether the speaker actually knows the secret):"
        : "泄密机检命中（请核实说话人是否知情）：",
    );
    for (const hit of params.leaks) {
      lines.push(
        `- [${hit.gapId}] "${hit.keyword}" ×${hit.occurrences}${hit.inDialogue ? (language === "en" ? " (in dialogue)" : "（对白内）") : ""}${language === "en" ? "" : `；知情人：${hit.knows.join("、") || "无"}`}`,
      );
    }
  }
  if (params.redundancies.length > 0) {
    lines.push(
      language === "en"
        ? "Reader-redundancy pre-scan flagged (reader already knows these):"
        : "废笔机检命中（读者已知，勿复述）：",
    );
    for (const hit of params.redundancies) {
      lines.push(`- [${hit.gapId}] "${hit.keyword}" ×${hit.occurrences}`);
    }
    void language;
  }
  return lines.join("\n");
}

File: engine-rs/src/utils/info_gap_ledger.rs
//! 信息差账本（G7a/341 号，Phase B 批次二）。
//!
//! TS 真源：`packages/core/src/utils/info-gap-ledger.ts`；共享向量：
//! `packages/core/src/__tests__/golden/info-gap-vectors.json`
//! （差分测试 `tests/golden_info_gap_diff.rs`）。
//!
//! truth 类目 `story/info_gaps.md`（作者可手改）+ 泄密机检 / 废笔机检
//! （continuity 审计维度 38/39 的本地预扫描）。

use serde::Deserialize;
use serde::Serialize;

pub const DIALOGUE_WINDOW_CHARS: usize = 24;

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InfoGapEntry {
    pub id: String,
    #[serde(default)]
    pub secret: String,
    #[serde(default)]
    pub knows: Vec<String>,
    #[serde(default)]
    pub reader_knows: bool,
    #[serde(default)]
    pub registered_at: i64,
    #[serde(default)]
    pub keywords: Vec<String>,
}

const GAP_HEADING: &str = "## gap-";

/// 解析信息差账本 markdown（作者手改容错：缺字段用空值，坏节跳过）。
pub fn parse_info_gaps_markdown(markdown: &str) -> Vec<InfoGapEntry> {
    let mut entries: Vec<InfoGapEntry> = Vec::new();
    let mut current_id: Option<String> = None;
    let mut fields: std::collections::HashMap<String, String> = std::collections::HashMap::new();
    let mut has_current = false;

    macro_rules! flush {
        () => {
            if has_current {
                let secret = fields.get("secret").map(|s| s.trim().to_string()).unwrap_or_default();
                let knows = split_list(fields.get("knows").map(String::as_str).unwrap_or(""));
                let reader_knows = matches!(
                    fields.get("readerKnows").map(String::as_str).unwrap_or("").trim().to_lowercase().as_str(),
                    "true" | "是" | "yes" | "1"
                );
                let registered_at = fields
                    .get("registeredAt")
                    .map(String::as_str)
                    .unwrap_or("")
                    .trim()
                    .parse::<i64>()
                    .unwrap_or(0);
                let keywords = split_list(fields.get("keywords").map(String::as_str).unwrap_or(""));
                if !secret.is_empty() || !keywords.is_empty() {
                    entries.push(InfoGapEntry {
                        id: current_id.clone().unwrap_or_default(),
                        secret,
                        knows,
                        reader_knows,
                        registered_at,
                        keywords,
                    });
                }
            }
            has_current = false;
        };
    }

    for raw_line in markdown.lines() {
        let line = raw_line.trim();
        if let Some(rest) = line.strip_prefix("## ") {
            let name = rest.trim();
            if name.starts_with(GAP_HEADING.trim_start_matches("## ")) && name.len() >= 5 {
                flush!();
                current_id = Some(name.to_string());
                has_current = true;
                fields.clear();
                continue;
            }
        }
        if !has_current {
            continue;
        }
        if let Some(rest) = line.strip_prefix("- ") {
            if let Some(colon) = rest.find(':') {
                let key = rest[..colon].trim().to_string();
                let value = rest[colon + 1..].trim_start().to_string();
                fields.insert(key, value);
            }
        }
    }
    flush!();
    entries
}

fn split_list(value: &str) -> Vec<String> {
    value
        .split([',', '，', '、', ';', '；'])
        .map(|item| item.trim().to_string())
        .filter(|item| !item.is_empty())
        .collect()
}

/// 渲染信息差账本（防呆方言：每条一节、平铺 key: value 行）。
pub fn render_info_gaps_markdown(entries: &[InfoGapEntry]) -> String {
    let sections: Vec<String> = entries
        .iter()
        .map(|entry| {
            format!(
                "## {}\n- secret: {}\n- knows: {}\n- readerKnows: {}\n- registeredAt: {}\n- keywords: {}",
                entry.id,
                entry.secret,
                entry.knows.join(", "),
                entry.reader_knows,
                entry.registered_at,
                entry.keywords.join(", "),
            )
        })
        .collect();
    let mut parts = vec!["# 信息差账本（Info Gaps）".to_string()];
    parts.extend(sections);
    parts.join("\n\n")
}

// ── 机检 ──

fn has_dialogue_mark(text: &str) -> bool {
    text.chars().any(|c| "「」『』“”‘’\"'".contains(c))
}

fn count_occurrences(content: &str, keyword: &str) -> u32 {
    if keyword.is_empty() {
        return 0;
    }
    let lower = content.to_lowercase();
    let needle = keyword.to_lowercase();
    let mut count = 0u32;
    let mut index = 0usize;
    while let Some(found) = lower[index..].find(&needle) {
        count += 1;
        index += found + needle.len();
    }
    count
}

/// 命中窗口（前后 window 字）是否含对白引号。
fn window_has_dialogue(content: &str, keyword: &str, window: usize) -> bool {
    let lower = content.to_lowercase();
    let needle = keyword.to_lowercase();
    let needle_chars = keyword.chars().count();
    let mut index = 0usize;
    while let Some(found) = lower[index..].find(&needle) {
        let byte_start = index + found;
        // 字节区间近似 TS 的码元窗口：以字符为单位计算前后窗。
        let char_start = lower[..byte_start].chars().count();
        let window_from = char_start.saturating_sub(window);
        let window_to = char_start + needle_chars + window;
        let window_text: String = content
            .chars()
            .skip(window_from)
            .take(window_to - window_from)
            .collect();
        if has_dialogue_mark(&window_text) {
            return true;
        }
        index += found + needle.len();
    }
    false
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecretLeakHit {
    pub gap_id: String,
    pub keyword: String,
    pub occurrences: u32,
    pub in_dialogue: bool,
    pub knows: Vec<String>,
}

/// 泄密机检：readerKnows=false 且已登记的秘密关键词命中即候选。
pub fn detect_secret_leaks(
    content: &str,
    gaps: &[InfoGapEntry],
    current_chapter: i64,
) -> Vec<SecretLeakHit> {
    let mut hits: Vec<SecretLeakHit> = Vec::new();
    for gap in gaps {
        if gap.reader_knows || gap.registered_at > current_chapter {
            continue;
        }
        for keyword in &gap.keywords {
            if keyword.is_empty() {
                continue;
            }
            let occurrences = count_occurrences(content, keyword);
            if occurrences == 0 {
                continue;
            }
            let in_dialogue = window_has_dialogue(content, keyword, DIALOGUE_WINDOW_CHARS);
            hits.push(SecretLeakHit {
                gap_id: gap.id.clone(),
                keyword: keyword.clone(),
                occurrences,
                in_dialogue,
                knows: gap.knows.clone(),
            });
        }
    }
    hits.sort_by(|a, b| {
        b.in_dialogue
            .cmp(&a.in_dialogue)
            .then(b.occurrences.cmp(&a.occurrences))
            .then(a.gap_id.cmp(&b.gap_id))
            .then(a.keyword.cmp(&b.keyword))
    });
    hits
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderRedundancyHit {
    pub gap_id: String,
    pub keyword: String,
    pub occurrences: u32,
}

/// 废笔机检：读者已知的秘密关键词再次出现——向读者复述已知信息。
pub fn detect_reader_redundancy(
    content: &str,
    gaps: &[InfoGapEntry],
) -> Vec<ReaderRedundancyHit> {
    let mut hits: Vec<ReaderRedundancyHit> = Vec::new();
    for gap in gaps {
        if !gap.reader_knows {
            continue;
        }
        for keyword in &gap.keywords {
            if keyword.is_empty() {
                continue;
            }
            let occurrences = count_occurrences(content, keyword);
            if occurrences > 0 {
                hits.push(ReaderRedundancyHit {
                    gap_id: gap.id.clone(),
                    keyword: keyword.clone(),
                    occurrences,
                });
            }
        }
    }
    hits.sort_by(|a, b| {
        b.occurrences
            .cmp(&a.occurrences)
            .then(a.gap_id.cmp(&b.gap_id))
            .then(a.keyword.cmp(&b.keyword))
    });
    hits
}

/// 审计注入文本：continuity 维度 38/39 的本地机检摘要（无命中返回 None）。
pub fn render_info_gap_audit_notes(
    leaks: &[SecretLeakHit],
    redundancies: &[ReaderRedundancyHit],
    language: Option<&str>,
) -> Option<String> {
    if leaks.is_empty() && redundancies.is_empty() {
        return None;
    }
    let is_en = language == Some("en");
    let mut lines: Vec<String> = Vec::new();
    if !leaks.is_empty() {
        lines.push(
            if is_en {
                "Secret-leak pre-scan flagged (verify whether the speaker actually knows the secret):"
            } else {
                "泄密机检命中（请核实说话人是否知情）："
            }
            .to_string(),
        );
        for hit in leaks {
            let dialogue_note = if hit.in_dialogue {
                if is_en { " (in dialogue)".to_string() } else { "（对白内）".to_string() }
            } else {
                String::new()
            };
            let knows_note = if is_en {
                String::new()
            } else if hit.knows.is_empty() {
                "；知情人：无".to_string()
            } else {
                format!("；知情人：{}", hit.knows.join("、"))
            };
            lines.push(format!(
                "- [{}] \"{}\" ×{}{}{}",
                hit.gap_id, hit.keyword, hit.occurrences, dialogue_note, knows_note
            ));
        }
    }
    if !redundancies.is_empty() {
        lines.push(
            if is_en {
                "Reader-redundancy pre-scan flagged (reader already knows these):"
            } else {
                "废笔机检命中（读者已知，勿复述）："
            }
            .to_string(),
        );
        for hit in redundancies {
            lines.push(format!("- [{}] \"{}\" ×{}", hit.gap_id, hit.keyword, hit.occurrences));
        }
    }
    Some(lines.join("\n"))
}

/// 机器可读契约（双端 golden 锁形状）。
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InfoGapContract {
    pub dimensions: Vec<InfoGapDimension>,
    pub truth_file: &'static str,
    pub dialogue_window: usize,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct InfoGapDimension {
    pub id: u32,
    pub zh: String,
    pub en: String,
}

pub fn info_gap_contract() -> InfoGapContract {
    InfoGapContract {
        dimensions: vec![
            InfoGapDimension { id: 38, zh: "泄密机检".into(), en: "Secret Leak Check".into() },
            InfoGapDimension { id: 39, zh: "废笔机检".into(), en: "Reader Redundancy Check".into() },
        ],
        truth_file: "story/info_gaps.md",
        dialogue_window: DIALOGUE_WINDOW_CHARS,
    }
}
